python/rascaline/rascaline/utils/atomic_density.py

```diff
@@ -0,0 +1,166 @@
+        atomic_gaussian_width_sq = self.atomic_gaussian_width**2
+        return np.exp(-0.5 * positions**2 / atomic_gaussian_width_sq) / (
+            np.pi * atomic_gaussian_width_sq
+        ) ** (3 / 4)
+
+
+class LodeDensity(AtomicDensityBase):
+    r"""Smeared Power Law Densities/
+
+    It is defined as
+
+    .. math::
+
+        g(r) = \frac{1}{\Gamma\left(\frac{p}{2}\right)}
+               \frac{\gamma\left( \frac{p}{2}, \frac{r^2}{2\sigma^2} \right)}
+                    {r^p},
+
+    where :math:`\Gamma(z)` is the Gamma function and :math:`\gamma(a, x)` is the
+    incomplete lower Gamma function. However its evaluation at :math:`r=0` is
+    problematic because :math:`g(r)` is of the form :math:`0/0`. For practical
+    implementations, it is thus more convenient to rewrite the density as
+
+    .. math::
+
+        g(r) = \frac{1}{\Gamma(a)}\frac{1}{\left(2 \sigma^2\right)^a}
+                \begin{cases}
+                    \frac{1}{a} - \frac{x}{a+1} + \frac{x^2}{2(a+2)} + \mathcal{O}(x^3)
+                        & x < 10^{-5} \\
+                    \frac{\gamma(a,x)}{x^a}
+                        & x \geq 10^{-5}
+                \end{cases}
+
+    It is convenient to use the expression for sufficiently small :math:`x` since the
+    relative weight of the first neglected term is on the order of :math:`1/6x^3`.
+    Therefore, the threshold :math:`x = 10^{-5}` leads to relative errors on the order
+    of the machine epsilon.
+
+    :param atomic_gaussian_width: Width of the atom-centered gaussian used to create the
+        atomic density
+    :param potential_exponent: Potential exponent of the decorated atom density.
+        Currently only implemented for potential_exponent < 10. Some exponents can be
+        connected to SOAP or physics-based quantities: p=0 uses Gaussian densities as in
+        SOAP, p=1 uses 1/r Coulomb like densities, p=6 uses 1/r^6 dispersion like
+        densities.
+    """
+
+    def __init__(self, atomic_gaussian_width: float, potential_exponent: int):
+        if not HAS_SCIPY:
+            raise ValueError("LodeDensity requires scipy!")
+
+        self.potential_exponent = potential_exponent
+        self.atomic_gaussian_width = atomic_gaussian_width
+
+    def _f_sr(self, a, x):
+        return 1 / a - x / (a + 1) + x**2 / (2 * (a + 2))
+
+    def _f_lr(self, a, x):
+        return gamma(a) * gammainc(a, x) / x**a
+
+    def compute(self, positions: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
+        if self.potential_exponent == 0:
+            return GaussianDensity.compute(self, positions=positions)
+        else:
+            atomic_gaussian_width_sq = self.atomic_gaussian_width**2
+            a = self.potential_exponent / 2
+            x = positions**2 / (2 * atomic_gaussian_width_sq)
+
+            prefac = 1 / gamma(a) / (2 * atomic_gaussian_width_sq) ** a
+
+            return prefac * np.where(
```

Comment:
Could you explain why you use a different implementation for short and long range here? And how the 1e-5 was picked?

Reply:
Because if you do not do this you get `0/0` terms. I tried to explain this in docstring.